cmd/query/app/apiv3/grpc_gateway_test.go

```diff
@@ -0,0 +1,100 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apiv3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"github.com/jaegertracing/jaeger/cmd/query/app/querysvc"
+	"github.com/jaegertracing/jaeger/model"
+	_ "github.com/jaegertracing/jaeger/pkg/gogocodec" //force gogo codec registration
+	"github.com/jaegertracing/jaeger/proto-gen/api_v3"
+	dependencyStoreMocks "github.com/jaegertracing/jaeger/storage/dependencystore/mocks"
+	spanstoremocks "github.com/jaegertracing/jaeger/storage/spanstore/mocks"
+)
+
+func TestGRPCGateway(t *testing.T) {
+	r := &spanstoremocks.Reader{}
+	traceID := model.NewTraceID(150, 160)
+	r.On("GetTrace", mock.AnythingOfType("*context.valueCtx"), mock.AnythingOfType("model.TraceID")).Return(
+		&model.Trace{
+			Spans: []*model.Span{
+				{
+					TraceID:       traceID,
+					SpanID:        model.NewSpanID(180),
+					OperationName: "foobar",
+				},
+			},
+		}, nil).Once()
+
+	q := querysvc.NewQueryService(r, &dependencyStoreMocks.Reader{}, querysvc.QueryServiceOptions{})
+	server := grpc.NewServer()
+	h := &Handler{
+		QueryService: q,
+	}
+	api_v3.RegisterQueryServiceServer(server, h)
+
+	lis, _ := net.Listen("tcp", ":0")
+	go func() {
+		err := server.Serve(lis)
+		require.NoError(t, err)
+	}()
+
+	router := &mux.Router{}
+	err := RegisterGRPCGateway(router, "", lis.Addr().String())
+	require.NoError(t, err)
+
+	httpLis, err := net.Listen("tcp", ":0")
+	require.NoError(t, err)
+	go func() {
+		err = http.Serve(httpLis, router)
+		require.NoError(t, err)
+		defer httpLis.Close()
+	}()
+	req, err := http.NewRequest("GET", fmt.Sprintf("http://localhost%s/v3/traces/123", strings.Replace(httpLis.Addr().String(), "[::]", "", 1)), nil)
+	req.Header.Set("Content-Type", "application/json")
+	response, err := http.DefaultClient.Do(req)
+	buf := bytes.Buffer{}
+	_, err = buf.ReadFrom(response.Body)
+	require.NoError(t, err)
+
+	jsonpb := &runtime.JSONPb{}
+	var envelope envelope
+	err = json.Unmarshal(buf.Bytes(), &envelope)
+	require.NoError(t, err)
+	var spansResponse api_v3.SpansResponseChunk
+	err = jsonpb.Unmarshal(envelope.Result, &spansResponse)
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(spansResponse.GetResourceSpans()))
+	assert.Equal(t, uint64ToTraceID(traceID.High, traceID.Low), spansResponse.GetResourceSpans()[0].GetInstrumentationLibrarySpans()[0].GetSpans()[0].GetTraceId())
+}
+
+// see https://github.com/grpc-ecosystem/grpc-gateway/issues/2189
```

Comment:
please add more color, is this a TODO that needs to be address in the future when the referenced ticket is solved?

Reply:
this is not TODO. I don't think they will resolve the issue. our IDL has some comments with workaround.

Reply:
ok, but still add some color because the purpose of this comment is unclear